### PR TITLE
Improve link_to with html content param

### PR DESCRIPTION
**Pull request by vyper**
Improve support for html content in link_to helper

# Before
```ruby
link_to(nil, routes.path(:posts)) do
  strong 'Post'
end
# => <a href="/posts/"><strong>Post</strong></a>
```

# Now
```ruby
link_to(routes.path(:posts)) do
  strong 'Post'
end
# => <a href="/posts/"><strong>Post</strong></a>
```

### CODE REVIEW

**Comment by AlfonsoUceda** [comment]
I like to me, what do you think @jodosha @joneslee85

**Comment by joneslee85** [comment]
I like it too but I am questioning about what if the user provide both the param and the block?

**Comment by AlfonsoUceda** [comment]
then I think it crash :S nice catch. @vyper please could you confirm this?

**Comment by vyper** [comment]
@AlfonsoUceda and @joneslee85: Really it would crash, but I changed implementation and I improved documentation.

**Comment by joneslee85** [comment]
@vyper :+1:

**Comment by jodosha** [comment]
@vyper I like the idea and thank you for this PR.
There is one behavior to change IMO, which is the one that accepts both `content` and `blk`. From my understanding the former gets ignored. This can be source of confusion for devs. We should instead detect and handle two cases with an `ArgumentError` exception.

  1. When both `content` and `blk` are specified (`link_to('foo', routes.path(:home)) { 'bar' }`)
  2. When none of them is specified (`link_to()`).

There is also a third case that can be confusing:

```ruby
<%= link_to('foo') { 'bar' } %>
```

In unit tests it returns:

```html
<a href="foo">
bar
</a>
```

Please note that's an invalid `href`attribute. If you specify `/foo`, our unit tests return the expected value, but if you use in combination with Lotus::View, its security feature will autoescape `/`.

Having such a liberal signature like `def link_to(*content, **args, &blk)` may lead to unexpected usages that can lead devs to confusion.

@vyper Can you please add tests for the behaviors above? At this point it's clear we also need to expand integration tests (see `test/integration/link_to_helper_test.rb`). And as a last thing, when we're sure that the tests cover these edge cases, please experiment with a different signature. It would be great to have a stricter version of it. Thank you very much.

**Comment by vyper** [comment]
@jodosha better that way?

File: lib/lotus/helpers/link_to_helper.rb [hunk 1]
```diff
@@ -9,17 +9,49 @@ module Helpers
...
+      #   end # => <a href="/users"><strong>Users</strong></a>
+      def link_to(*args, **options, &blk)
+        if block_given? && args.size != 1
+          raise ArgumentError, 'Only an uri is expected as an argument when a block is given'
+        end
+
+        if !block_given? && args.size != 2
+          raise ArgumentError, 'Both content and url are expected as arguments'
+        end
+
```
Comment (joneslee85) [inline lib/lotus/helpers/link_to_helper.rb]: what happen if there are more than 3 arguments?

Comment (vyper) [inline lib/lotus/helpers/link_to_helper.rb]: raises exception `Both content and url are expected as arguments` when without block and `Only an uri is expected as an argument when a block is given` with block.

File: test/link_to_helper_test.rb [hunk 2]
```diff
@@ -20,4 +20,32 @@
...
+
+  it 'returns a link with html content' do
+    @view.link_to_with_html_content.must_equal %(<a href="/posts/">\n<strong>Post</strong>\n</a>)
+  end
+
+  it 'returns a link with html content, id and class' do
+    @view.link_to_with_html_content_id_and_class.must_equal %(<a id="posts__link" class="first" href="/posts/">\n<strong>Post</strong>\n</a>)
+  end
+
+  it 'raises an link with content and html content' do
```
Comment (joneslee85) [inline test/link_to_helper_test.rb]: s/an/a

Comment (joneslee85) [inline test/link_to_helper_test.rb]: raises Error not a link

**Comment by joneslee85** [comment]
@vyper looks good now, please squash into one and I'll get it merge

**Comment by vyper** [comment]
@joneslee85 done. (:

**Comment by jodosha** [comment]
@vyper @joneslee85 I've implemented a stricter signature in https://github.com/lotus/helpers/commit/153b65977e1fde416ef18826c037f5cbd7f60bb4, by keeping the same features.

**Comment by joneslee85** [comment]
@jodosha your implementation :+1:
